tests/karate/src/test/java/be/utils/JsonConverter.java

```diff
@@ -0,0 +1,163 @@
+
+  /**
+   * Produces the base64 variant of the json file passed as argument
+   */
+  public String convertJsonToBase64(Json json){
+    String stringJson = json.toString();
+    byte[] jsonBytes = stringJson.getBytes();
+    Base64.Encoder encoder = Base64.getEncoder();
+    String base64 = encoder.encodeToString(jsonBytes);
+    return base64;
+  }
+
+  /**
+   * Produces a valid Json representation of a message given the message data, the id of the message
+   * and the channel where the message is supposed to be sent
+   */
+  public Json publishМessageFromData(String stringData, int id, String channel){
+    Json messageData = Json.of(stringData);
+    String messageDataBase64 = convertJsonToBase64(messageData);
+    Map<String,Object> messageJson = new LinkedHashMap<>();
+
+    messageJson.put("method", "publish");
+    messageJson.put("id", id);
+
+    Map<String,Object> paramsPart = constructParamsField(channel, stringData);
+
+    messageJson.put("params", paramsPart);
+    messageJson.put("jsonrpc", "2.0");
+
+    return Json.of(messageJson);
+  }
+
+  public Map<String,Object> constructMessageField(String stringData, boolean forcedMessageId){
+    Map<String,Object> messagePart  = new LinkedHashMap<>();
+    Json messageData = Json.of(stringData);
+    String messageDataBase64 = convertJsonToBase64(messageData);
+    messagePart.put("data", messageDataBase64);
+    messagePart.put("sender", senderPk);
+
+    String signature = constructSignature(stringData);
+    messagePart.put("signature", signature);
+    String messageId = hash(messageDataBase64.getBytes(),signature.getBytes());
+
+    if(forcedMessageId && !messageIdForced.isEmpty()){
+      messageId = this.messageIdForced;
+      messageIdForced = "";
+    }
+    messagePart.put("message_id", messageId);
+    System.out.println("message id is : " + messageId);
+
+    String[] witness = new String[0];
+    messagePart.put("witness_signatures", witness);
+
+    return messagePart;
+  }
+
+  public Map<String,Object> constructParamsField(String channel, String messageDataBase64){
+    Map <String,Object> paramsPart = new LinkedHashMap<>();
+    paramsPart.put("channel", channel);
+
+    Map<String,Object> messagePart = constructMessageField(messageDataBase64,false);
+    paramsPart.put("message", messagePart);
+
+    return paramsPart;
+  }
+
+  /**
+   * Constructs a valid signature on given data
+   */
+  public String constructSignature(String messageDataBase64){
+    try {
+      // Hex representation of the private key
+      byte[] privateKeyBytes = new byte[privateKeyHex.length() / 2];
+
+      for (int i = 0; i < privateKeyBytes.length; i++) {
+        int index = i * 2;
+
+        int val = Integer.parseInt(privateKeyHex.substring(index, index + 2), 16);
+        privateKeyBytes[i] = (byte)val;
+      }
+      PublicKeySign publicKeySign = new Ed25519Sign(privateKeyBytes);
+      byte[] singie = publicKeySign.sign(messageDataBase64.getBytes());
```

Comment:
This was correct actually. Because messageDataBase64 is not not actually in base64 in your implementation.

It is simply the stringData which makes this line correct.

(Although defining the charset as UTF-8 was good so you should put it back 😅)

But you definitely need to rename this field as the current name is definitely missleading.

Reply:
I agree on this point, and applied the changes